components/dada-validate/src/validate/validator.rs

```diff
@@ -767,6 +760,124 @@ impl<'me> Validator<'me> {
         self.exit(validated_expr)
     }
 
+    ///
+    fn concatenate(
+        &mut self,
+        concatenate_expr: syntax::Expr,
+        exprs: &[syntax::Expr],
+    ) -> validated::Expr {
+        let margin = self.compute_margin(exprs);
+
+        let mut validated_exprs = Vec::with_capacity(exprs.len());
+        for (expr, index) in exprs.iter().zip(0..) {
+            if let syntax::ExprData::StringLiteral(word) = expr.data(self.syntax_tables()) {
+                let word_str = word.as_str(self.db);
+                let without_margin = self.strip_margin(margin, word_str);
+                let mut without_margin = &without_margin[..];
+
+                // We always strip leading/trailing newline, because people tend to write
+                // things like
+                //
+                // ```
+                // print("
+                //     Foo
+                // ");
+                // ```
+                //
+                // and we want that to print "Foo". Note that we haven't applied
+                // escapes yet, so `print("\n")` is unaffected.
+                if index == 0 {
+                    if let Some(s) = without_margin.strip_prefix('\n') {
+                        without_margin = s;
+                    }
+                } else if index == exprs.len() - 1 {
+                    if let Some(s) = without_margin.strip_suffix('\n') {
+                        without_margin = s;
```

Comment:
When will we get here? I can't find a case. I think the suffix `\n` is always striped in `strip_margin`.

Reply:
Oh, I found a case
```
async fn main() {
    print("
    hello
    {"world"}
    ").await
}
```
But it does not strip trailing newline for a string literal, like the documented example, because the `index` is 0 and we don't get into this condition branch.
```
async fn main() {
  print("
  foo
  ").await
}
```

Reply:
Hmm, yeah, something is broken with this logic